docs/source/format/Columnar.rst

```diff
@@ -487,6 +499,102 @@ will be represented as follows: ::
           |-------------------------------|-----------------------|
           | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 | unspecified (padding) |
 
+ListView Layout
+~~~~~~~~~~~~~~~
+
+The ListView layout is defined by three buffers: a validity bitmap, an offsets
+buffer, and an additional sizes buffer. Sizes and offsets have the identical bit
+width and both 32-bit and 64-bit signed integer options are supported.
+
+As in the List layout, the offsets encode the start position of each slot in the
+child array. In contrast to the List layout, list lengths are stored explicitly
+in the sizes buffer instead of inferred. This allows offsets to be out of order.
+Elements of the child array do not have to be stored in the same order they
+logically appear in the list elements of the parent array.
+
+When a value is null, the corresponding offset and size can have arbitrary
+values. When size is 0, the corresponding offset can have an arbitrary value.
+If choosing a value is possible, we recommend setting offsets and sizes to 0 in
```

Comment:
I'm curious why there would be situations where picking a value is not possible?

Reply:
It's not so much about possibility (so I'm very open to re-wording), but more about machine affinity (having less branches in loops, SIMD...). I can't predict all the possible ways people might want to save time in compute kernels.

For instance, when I slice the list-views for concatenation I have to subtract values from offsets (because I'm compressing the child array). I can do it in multiple ways:

 - `offsets[i] = offsets[i] - displacement;` leading to the possibility of negative offsets when `sizes[i] == 0` and `offsets[i]` is `unspecified` and potentially smaller than the positive displacement.
 - `offsets[i] = std::max(0, offsets[i] - displacement);` guaranteeing it's zero but introducing branching `*`
 - skipping setting `offsets[i]` at all when `sizes[i] == 0` introducing branching
 
I don't think this is an important enough operation but it illustrates how easy these stricter requirements could get in the way of micro optimizations. I suspect these become even more relevant when attempts to vectorize these loops are made.

`*` or requiring branchless saturating math operations

Reply:
In my experience forcing kernels to handle the null mask often results in significantly less performant code, especially in the absence of hand-rolled SIMD, I suspect this will outweigh any potential benefits from allowing arbitrary values to be left at null positions.

In your example, surely you have to perform the bounds checks anyway as the non-null elements might be shorted than displacement?

Reply:
> In your example, surely you have to perform the bounds checks anyway as the non-null elements might be shorted than displacement?

`displacement` is greater than `0` but `<=` all the non-null and non-empty list-view offsets in the buffer. It's calculated by finding the smallest valid offset on the source array before applying the transformation. Finding the smallest requires checking null mask and `sizes[i]` buffer, applying the transformation is where many alternatives are available.

I will check the C++ implementation again, but in the Go implementation I'm doing all the checks.

Being more strict on the spec is more likely to constrain kernels writing/transforming list-views, but hopefully we compensate for that in efficiency of kernels reading list-views.

Reply:
In systems that also support selection vectors and not only validity bitmaps, skipping a pass over the values becomes a powerful optimization. Even more important than being able to blindly read or transform values at the null slots.

https://dl.acm.org/doi/abs/10.1145/3465998.3466009

Reply:
Agreed, many kernels in arrow-rs do still handle the null mask, or some similar mask derived from the inputs, but make use CLZ tricks to iterate it efficiently. This in turn requires providing an API for value access, that doesn't itself then check the null mask, and in order for this to be safe we need the invariants to hold for non-null values :smile: